libkbfs/tlf_journal.go

```diff
@@ -0,0 +1,611 @@
+	delegateBlockServer BlockServer
+	log                 logger.Logger
+	deferLog            logger.Logger
+
+	// All the channels below are used as simple on/off
+	// signals. They're buffered for one object, and all sends are
+	// asynchronous, so multiple sends get collapsed into one
+	// signal.
+	hasWorkCh      chan struct{}
+	needPauseCh    chan struct{}
+	needResumeCh   chan struct{}
+	needShutdownCh chan struct{}
+
+	// Protects all operations on blockJournal and mdJournal.
+	//
+	// TODO: Don't let flushing block put operations. See
+	// KBFS-1433.
+	//
+	// TODO: Consider using https://github.com/pkg/singlefile
+	// instead.
+	lock sync.RWMutex
+
+	blockJournal *blockJournal
+	mdJournal    *mdJournal
+
+	bwDelegate tlfJournalBWDelegate
+}
+
+func makeTLFJournal(
+	ctx context.Context, dir string, tlfID TlfID, config tlfJournalConfig,
+	delegateBlockServer BlockServer, log logger.Logger,
+	bws TLFJournalBackgroundWorkStatus,
+	bwDelegate tlfJournalBWDelegate) (*tlfJournal, error) {
+	tlfDir := filepath.Join(dir, tlfID.String())
+
+	blockJournal, err := makeBlockJournal(
+		ctx, config.Codec(), config.Crypto(), tlfDir, log)
+	if err != nil {
+		return nil, err
+	}
+
+	_, uid, err := config.KBPKI().GetCurrentUserInfo(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	key, err := config.KBPKI().GetCurrentVerifyingKey(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	mdJournal, err := makeMDJournal(
+		uid, key, config.Codec(), config.Crypto(), tlfDir, log)
+	if err != nil {
+		return nil, err
+	}
+
+	j := &tlfJournal{
+		tlfID:               tlfID,
+		config:              config,
+		delegateBlockServer: delegateBlockServer,
+		log:                 log,
+		deferLog:            log.CloneWithAddedDepth(1),
+		hasWorkCh:           make(chan struct{}, 1),
+		needPauseCh:         make(chan struct{}, 1),
+		needResumeCh:        make(chan struct{}, 1),
+		needShutdownCh:      make(chan struct{}, 1),
+		blockJournal:        blockJournal,
+		mdJournal:           mdJournal,
+		bwDelegate:          bwDelegate,
+	}
+
+	go j.doBackgroundWorkLoop(bws)
+
+	// Signal work to pick up any existing journal entries.
+	select {
+	case j.hasWorkCh <- struct{}{}:
+	default:
+	}
+
+	j.log.CDebugf(ctx, "Enabled journal for %s with path %s", tlfID, tlfDir)
+	return j, nil
+}
+
+// doBackgroundWorkLoop is the main function for the background
+// goroutine. It just calls doBackgroundWork whenever there is work.
+func (j *tlfJournal) doBackgroundWorkLoop(bws TLFJournalBackgroundWorkStatus) {
+	ctx := context.Background()
+	if j.bwDelegate != nil {
+		ctx = j.bwDelegate.GetBackgroundContext()
+	}
+	ctx = ctxWithRandomID(ctx, "journal-auto-flush", "1", j.log)
+	defer func() {
+		if j.bwDelegate != nil {
+			j.bwDelegate.OnShutdown(ctx)
+		}
+	}()
+	// errCh and bwCancel are non-nil only when bws ==
+	// TLFJournalBackgroundWorkEnabled and there's work being done
+	// in the background.
+	var errCh <-chan error
+	var bwCancel context.CancelFunc
+	for {
+		switch {
```

Comment:
I thought about that, but I'd rather have all the states be on the same indent level. Plus, it's already so indented. :(